File: jas_ocaml/lib/geometry/path_text_layout.ml
(** Path text layout. See [path_text_layout.mli]. *)

type path_glyph = {
  idx : int;
  offset : float;
  width : float;
  cx : float;
  cy : float;
  angle : float;
  overflow : bool;
}

type t = {
  glyphs : path_glyph array;
  total_length : float;
  font_size : float;
  char_count : int;
}

let arc_lengths pts =
  let rec go acc prev = function
    | [] -> List.rev acc
    | (x, y) :: rest ->
      let (px, py) = prev in
      let dx = x -. px in
      let dy = y -. py in
      let len = (List.hd acc) +. sqrt (dx *. dx +. dy *. dy) in
      go (len :: acc) (x, y) rest
  in
  match pts with
  | [] -> [0.0]
  | first :: rest -> go [0.0] first rest

let sample_at_arc pts_arr len_arr arc =
  let n = Array.length pts_arr in
  if n < 2 then
    let p = if n = 1 then pts_arr.(0) else (0.0, 0.0) in
    let (x, y) = p in (x, y, 0.0)
  else begin
    let arc = max 0.0 arc in
    let result = ref None in
    (try
      for i = 1 to n - 1 do
        if len_arr.(i) >= arc then begin
          let seg = len_arr.(i) -. len_arr.(i - 1) in
          let t = if seg > 0.0 then (arc -. len_arr.(i - 1)) /. seg else 0.0 in
          let (ax, ay) = pts_arr.(i - 1) in
          let (bx, by) = pts_arr.(i) in
          let x = ax +. t *. (bx -. ax) in
          let y = ay +. t *. (by -. ay) in
          let angle = atan2 (by -. ay) (bx -. ax) in
          result := Some (x, y, angle);
          raise Exit
        end
      done
    with Exit -> ());
    match !result with
    | Some r -> r
    | None ->
      let last = n - 1 in
      let (ax, ay) = pts_arr.(last - 1) in
      let (bx, by) = pts_arr.(last) in
      (bx, by, atan2 (by -. ay) (bx -. ax))
  end

let layout d content start_offset font_size measure =
  let pts = Element.flatten_path_commands d in
  let lengths = arc_lengths pts in
  let total = match List.rev lengths with [] -> 0.0 | x :: _ -> x in
  let n = String.length content in
  if total <= 0.0 || pts = [] then
    { glyphs = [||]; total_length = total; font_size; char_count = n }
  else begin
    let pts_arr = Array.of_list pts in
    let len_arr = Array.of_list lengths in
    let start_arc = (max 0.0 (min 1.0 start_offset)) *. total in
    let cur_arc = ref start_arc in
    let glyphs = ref [] in
    for i = 0 to n - 1 do
      let cw = measure (String.make 1 content.[i]) in
      let center_arc = !cur_arc +. cw /. 2.0 in
      let overflow = center_arc > total in
      let (cx, cy, angle) = sample_at_arc pts_arr len_arr (min center_arc total) in
      glyphs := { idx = i; offset = !cur_arc; width = cw; cx; cy; angle; overflow } :: !glyphs;
      cur_arc := !cur_arc +. cw
    done;
    { glyphs = Array.of_list (List.rev !glyphs);
      total_length = total; font_size; char_count = n }
  end

let cursor_pos t cursor =
  let n = Array.length t.glyphs in
  if n = 0 then None
  else if cursor = 0 then begin
    let g = t.glyphs.(0) in
    let dx = -. cos g.angle *. g.width /. 2.0 in
    let dy = -. sin g.angle *. g.width /. 2.0 in
    Some (g.cx +. dx, g.cy +. dy, g.angle)
  end else if cursor >= n then begin
    let g = t.glyphs.(n - 1) in
    let dx = cos g.angle *. g.width /. 2.0 in
    let dy = sin g.angle *. g.width /. 2.0 in
    Some (g.cx +. dx, g.cy +. dy, g.angle)
  end else begin
    let g = t.glyphs.(cursor) in
    let dx = -. cos g.angle *. g.width /. 2.0 in
    let dy = -. sin g.angle *. g.width /. 2.0 in
    Some (g.cx +. dx, g.cy +. dy, g.angle)
  end

let hit_test t x y =
  let n = Array.length t.glyphs in
  if n = 0 then 0
  else begin
    let best_idx = ref 0 in
    let best_dist = ref infinity in
    for i = 0 to n - 1 do
      let g = t.glyphs.(i) in
      let half = g.width /. 2.0 in
      let bx = g.cx -. cos g.angle *. half in
      let by = g.cy -. sin g.angle *. half in
      let ax = g.cx +. cos g.angle *. half in
      let ay = g.cy +. sin g.angle *. half in
      let db = sqrt ((x -. bx) *. (x -. bx) +. (y -. by) *. (y -. by)) in
      let da = sqrt ((x -. ax) *. (x -. ax) +. (y -. ay) *. (y -. ay)) in
      if db < !best_dist then begin best_dist := db; best_idx := i end;
      if da < !best_dist then begin best_dist := da; best_idx := i + 1 end
    done;
    !best_idx
  end

File: JasSwift/Sources/Algorithms/PathTextLayout.swift
import Foundation

// MARK: - Path text layout
//
// Walks the arc length of a flattened path and places one glyph per
// character at the center of its arc-length interval. Mirrors
// `path_text_layout.rs` / `.ml` / `.py`.

public struct PathGlyph {
    public let idx: Int
    public let offset: Double
    public let width: Double
    public let cx: Double
    public let cy: Double
    public let angle: Double
    public let overflow: Bool
}

public struct PathTextLayout {
    public let glyphs: [PathGlyph]
    public let totalLength: Double
    public let fontSize: Double
    public let charCount: Int
}

private func arcLengths(_ pts: [(Double, Double)]) -> [Double] {
    if pts.isEmpty { return [0.0] }
    var out: [Double] = [0.0]
    var prev = pts[0]
    for i in 1..<pts.count {
        let (x, y) = pts[i]
        let dx = x - prev.0, dy = y - prev.1
        out.append(out.last! + (dx * dx + dy * dy).squareRoot())
        prev = pts[i]
    }
    return out
}

private func sampleAtArc(_ pts: [(Double, Double)], _ lens: [Double], _ arcIn: Double)
    -> (Double, Double, Double) {
    let n = pts.count
    if n < 2 {
        let (x, y) = n == 1 ? pts[0] : (0.0, 0.0)
        return (x, y, 0.0)
    }
    let arc = max(0.0, arcIn)
    for i in 1..<n {
        if lens[i] >= arc {
            let seg = lens[i] - lens[i - 1]
            let t = seg > 0 ? (arc - lens[i - 1]) / seg : 0.0
            let (ax, ay) = pts[i - 1]
            let (bx, by) = pts[i]
            return (ax + t * (bx - ax), ay + t * (by - ay), atan2(by - ay, bx - ax))
        }
    }
    let last = n - 1
    let (ax, ay) = pts[last - 1]
    let (bx, by) = pts[last]
    return (bx, by, atan2(by - ay, bx - ax))
}

/// Layout `content` along the curve described by `d`, starting at the
/// fractional `startOffset` (0…1) along the path's arc length.
public func layoutPathText(_ d: [PathCommand],
                           content: String,
                           startOffset: Double,
                           fontSize: Double,
                           measure: (String) -> Double) -> PathTextLayout {
    let pts = flattenPathCommands(d)
    let lens = arcLengths(pts)
    let total = lens.last ?? 0.0
    let chars = Array(content)
    let n = chars.count
    if total <= 0 || pts.isEmpty {
        return PathTextLayout(glyphs: [], totalLength: total, fontSize: fontSize, charCount: n)
    }
    let startArc = max(0.0, min(1.0, startOffset)) * total
    var curArc = startArc
    var glyphs: [PathGlyph] = []
    for i in 0..<n {
        let cw = measure(String(chars[i]))
        let centerArc = curArc + cw / 2
        let overflow = centerArc > total
        let (cx, cy, angle) = sampleAtArc(pts, lens, min(centerArc, total))
        glyphs.append(PathGlyph(idx: i, offset: curArc, width: cw,
                                cx: cx, cy: cy, angle: angle, overflow: overflow))
        curArc += cw
    }
    return PathTextLayout(glyphs: glyphs, totalLength: total, fontSize: fontSize, charCount: n)
}

public extension PathTextLayout {
    /// Caret position (x, y, tangent angle) for `cursor`. nil if empty.
    func cursorPos(_ cursor: Int) -> (Double, Double, Double)? {
        let n = glyphs.count
        if n == 0 { return nil }
        if cursor == 0 {
            let g = glyphs[0]
            let dx = -cos(g.angle) * g.width / 2
            let dy = -sin(g.angle) * g.width / 2
            return (g.cx + dx, g.cy + dy, g.angle)
        }
        if cursor >= n {
            let g = glyphs[n - 1]
            let dx = cos(g.angle) * g.width / 2
            let dy = sin(g.angle) * g.width / 2
            return (g.cx + dx, g.cy + dy, g.angle)
        }
        let g = glyphs[cursor]
        let dx = -cos(g.angle) * g.width / 2
        let dy = -sin(g.angle) * g.width / 2
        return (g.cx + dx, g.cy + dy, g.angle)
    }

    func hitTest(_ x: Double, _ y: Double) -> Int {
        let n = glyphs.count
        if n == 0 { return 0 }
        var bestIdx = 0
        var bestDist = Double.infinity
        for i in 0..<n {
            let g = glyphs[i]
            let half = g.width / 2
            let bx = g.cx - cos(g.angle) * half
            let by = g.cy - sin(g.angle) * half
            let ax = g.cx + cos(g.angle) * half
            let ay = g.cy + sin(g.angle) * half
            let db = ((x - bx) * (x - bx) + (y - by) * (y - by)).squareRoot()
            let da = ((x - ax) * (x - ax) + (y - ay) * (y - ay)).squareRoot()
            if db < bestDist { bestDist = db; bestIdx = i }
            if da < bestDist { bestDist = da; bestIdx = i + 1 }
        }
        return bestIdx
    }
}
